block-layer/codecs/encrypted.md

```diff
@@ -0,0 +1,83 @@
+# Specification: Encrypted Block Codec
+
+**Status: Prescriptive - Draft**
+
+This document describes codecs for IPLD blocks (CID + Data) that are encrypted. The
+multicodec idenfier for the cipher and the initital vector are included in the block
+format and parsed into a standardized data model representation.
+
+The following known ciphers may be referenced in encrypted blocks:
+
+| name | multicodec |
+| --- | --- |
+| aes-gcm | 0x1401 |
+| aes-cbc | 0x1402 |
+| aes-ctr | 0x1403 |
+
+## What this spec is not
+
+This is not a complete system for application privacy. The following issues are
+out of scope for this specification, although they can obviously leverage these codecs:
+
+* Key signaling
+* Access controls
+* Dual-layer encryption w/ replication keys
+
+How you determine what key to apply to an encrypted block will need to be done in the
+application layer. How you decide to encrypt a graph and potentially link the encrypted
+blocks together for replication is done at the application layer. How you manage and access
+keys is done in the application layer.
+
+## Encode/Decode vs Encrypt/Decrypt
+
+The goal of specifying codecs that are used for encryption is to allow the codecs to
+include encryption and decryption programs alongside the codec. However, encrypting and
+decrypting are done by the user and are not done automatically as part of any encode/decode
+operation in the codec.
+
+The encryption program returns a data model value suitable for the block encode program. The
+decode program provides a data model value suitable for the decryption program. And the decryption
+program provides a data model value suitable for parsing into a new block (CID and Bytes). These
+programs are designed to interoperate but it's up to the user to combine them and provide the
+necessary key during encryption and decryption.
+
+## Encrypted Block Format
+
+An encrypted block can be decoded into its initializing vector and the encrypted byte
+payload. Since the initializing vector is a known size for each AES variant the block
+format is simply the iv followed by the byte payload.
+
+```
+| varint(cipher-multicodec) | varint(ivLength) | iv | bytes |
```

Comment:
Not that any justification is needed to propose a new block format, but what is the context for using this over cbor?

Is it because we think encryption is so useful and common that shaving the bytes off the field name is worthwhile? Is it because we think this format is so much simpler to implement than cbor that it can be used across more of the ecosystem? Something else?

Reply:
The most obvious reason is that it’s substantially smaller than CBOR. There’s no reason for us to leverage large general purpose block formats when jamming a few buffers and/or varints together will work.

We should never assume that CBOR or DagCBOR are already available. These formats are not *that* widely used outside of our bubble and are nowhere near the adoption level that I’d consider we can “take for granted” like we might JSON.

When we leverage an existing codec we then **also** have to do schema validation on the input and output in order to ensure determinism since the generic codec will allow any number of generic properties to be there in addition to what we’ve defined in the spec.

If you can write a codec in a few lines of code for a new data structure that is already its own multicodec identifier we shouldn’t shy away from doing just that. We avoid all the caveats and concerns about determinism when we write self-describing types as an ordered concatenation of bytes.

Reply:
yep, I figured 👍. The one caveat I have is:

> If you can write a codec in a few lines of code for a new data structure that is already its own multicodec identifier we shouldn’t shy away from doing just that.

I think we should still shy away from it since I'd prefer the number of codecs to be smaller rather than bigger as it puts pressure on generalized systems like IPFS to get bigger and bigger as they need to add support for more and more codecs. They also take up slots in the code table and arguably any block format could want a small code number because people could be storing tons of blocks and could then easily save some space.

Reply:
I think this is a good question.   "why would we prefer x over cbor?" is almost always a valid question.

In this case: I could see users remarking on the bytes shaved off as useful, yes.  And the simplicity also does seem potentially relevant: that this format will never, ever need to be capable of any kind of recursion, since it's just a small wrapper for the ciphertext, does make it possible to implement with considerably simpler mechanisms than cbor.

I don't think either of those arguments is open-and-shut, but they're enough to make me receptive to a new block format, at least.